clientlibs/js/src/ApiService/ApiService.ts

```diff
@@ -0,0 +1,311 @@
+import { UpGradeClientEnums, UpGradeClientInterfaces, UpGradeClientRequests } from '../types';
+import { DefaultHttpClient } from '../DefaultHttpClient/DefaultHttpClient';
+import {
+  CaliperEnvelope,
+  IExperimentAssignmentv5,
+  IFeatureFlag,
+  IFlagVariation,
+  IGroupMetric,
+  ILogInput,
+  ISingleMetric,
+} from 'upgrade_types';
+import { DataService } from 'DataService/DataService';
+import { IApiServiceRequestParams, IEndpoints } from './ApiService.types';
+import { IMarkDecisionPointParams } from 'UpGradeClient/UpGradeClient.types';
+import * as uuid from 'uuid';
+
+// this variable is used by webpack to replace the value of USE_CUSTOM_HTTP_CLIENT with true or false to create two different builds
+declare const USE_CUSTOM_HTTP_CLIENT: boolean;
+declare const IS_BROWSER: boolean;
+
+export default class ApiService {
+  private context: string;
+  private hostUrl: string;
+  private userId: string;
+  private token: string;
+  private apiVersion: string;
+  private clientSessionId: string;
+  private httpClient: UpGradeClientInterfaces.IHttpClientWrapper;
+  private api: IEndpoints;
+
+  constructor(config: UpGradeClientInterfaces.IConfig, private dataService: DataService) {
+    this.context = config.context;
+    this.hostUrl = config.hostURL;
+    this.token = config.token;
+    this.clientSessionId = config.clientSessionId;
+    this.userId = config.userId;
+    this.apiVersion = config.apiVersion;
+    this.api = {
+      init: `${this.hostUrl}/api/${this.apiVersion}/init`,
+      getAllExperimentConditions: `${this.hostUrl}/api/${this.apiVersion}/assign`,
+      markDecisionPoint: `${this.hostUrl}/api/${this.apiVersion}/mark`,
+      setGroupMemberShip: `${this.hostUrl}/api/${this.apiVersion}/groupmembership`,
+      setWorkingGroup: `${this.hostUrl}/api/${this.apiVersion}/workinggroup`,
+      failedExperimentPoint: `${this.hostUrl}/api/${this.apiVersion}/failed`,
+      getAllFeatureFlag: `${this.hostUrl}/api/${this.apiVersion}/featureflag`,
+      log: `${this.hostUrl}/api/${this.apiVersion}/log`,
+      logCaliper: `${this.hostUrl}/api/${this.apiVersion}/log/caliper`,
+      altUserIds: `${this.hostUrl}/api/${this.apiVersion}/useraliases`,
+      addMetrics: `${this.hostUrl}/api/${this.apiVersion}/metric`,
+    };
+    this.httpClient = this.setHttpClient(config.httpClient);
+  }
+
+  private setHttpClient(httpClient: UpGradeClientInterfaces.IHttpClientWrapper) {
+    console.log({ USE_CUSTOM_HTTP_CLIENT });
+    if (USE_CUSTOM_HTTP_CLIENT) {
+      if (httpClient) {
+        return httpClient;
+      } else {
+        throw new Error('Please provide valid httpClient.');
+      }
+    } else {
+      return new DefaultHttpClient();
```

Comment:
We could figure a way to bundle this in another library is there's client demand.  But including a basic lightweight implementation is probably fine.

Reply:
hm my goal was to find some webpack way to create two different bundles depending on the USE_CUSTOM_HTTP_CLIENT boolean supplied by webpack, one with the DefaultHttpClient and one without it (`browser` v `browser-lite`), but I deceived myself into thinking this would work. The package sizes are the same, so yeah it's bundling with.

I don't want the library user to have to import two packages, `upgrade-client-lib` and `upgrade-client-default-http-client`, just to have to marry them together in `new UpGradeClient('blah', 'blah', 'blah', { httpClient: new DefaultHttpClient() }})`... I'd the like "full" bundle to just have it by default without having to pass it in as a param, and the "lite" bundle to omit it and throw an error it is not passed in. Is there some simpler way to accomplish that I'm overthinking?

Reply:
Webpack configs are not my thing.  I've got no real advice to offer there.

Thinking about it differently, if you had the "other" package with the default client bundle everything together and the main package just have the basics.  Then, you're not using Webpack to carve anything out, you're using it to bring things together in the case of the "other" package.

Reply:
You can use webpack externals https://webpack.js.org/configuration/externals/ for removing the DefaultHttpClient if USE_CUSTOM_HTTP_CLIENT flag is true and given that axios is used only inside this client it will also get excluded from the bundle.

Reply:
great, I'll look into that